lightning/src/ln/wire.rs

```diff
@@ -0,0 +1,411 @@
+//! Wire encoding/decoding for Lightning messages according to [BOLT #1].
+//!
+//! Messages known by this module can be read from the wire using [`read`].
+//! The [`Message`] enum returned by [`read`] wraps the decoded message or the message type (if
+//! unknown) to use with pattern matching.
+//!
+//! Messages implementing the [`Encode`] trait define a message type and can be sent over the wire
+//! using [`write`].
+//!
+//! [BOLT #1]: https://github.com/lightningnetwork/lightning-rfc/blob/master/01-messaging.md
+//! [`read`]: fn.read.html
+//! [`write`]: fn.write.html
+//! [`Encode`]: trait.Encode.html
+//! [`Message`]: enum.Message.html
+
+use ln::msgs;
+use util::ser::{Readable, Writeable, Writer};
+
+/// A Lightning message returned by [`read`] when decoding bytes received over the wire. Each
+/// variant contains a message from [`ln::msgs`] or otherwise the message type if unknown.
+///
+/// [`read`]: fn.read.html
+/// [`ln::msgs`]: ../msgs/index.html
+#[allow(missing_docs)]
+pub enum Message {
+	Init(msgs::Init),
+	Error(msgs::ErrorMessage),
+	Ping(msgs::Ping),
+	Pong(msgs::Pong),
+	OpenChannel(msgs::OpenChannel),
+	AcceptChannel(msgs::AcceptChannel),
+	FundingCreated(msgs::FundingCreated),
+	FundingSigned(msgs::FundingSigned),
+	FundingLocked(msgs::FundingLocked),
+	Shutdown(msgs::Shutdown),
+	ClosingSigned(msgs::ClosingSigned),
+	UpdateAddHTLC(msgs::UpdateAddHTLC),
+	UpdateFulfillHTLC(msgs::UpdateFulfillHTLC),
+	UpdateFailHTLC(msgs::UpdateFailHTLC),
+	UpdateFailMalformedHTLC(msgs::UpdateFailMalformedHTLC),
+	CommitmentSigned(msgs::CommitmentSigned),
+	RevokeAndACK(msgs::RevokeAndACK),
+	UpdateFee(msgs::UpdateFee),
+	ChannelReestablish(msgs::ChannelReestablish),
+	AnnouncementSignatures(msgs::AnnouncementSignatures),
+	ChannelAnnouncement(msgs::ChannelAnnouncement),
+	NodeAnnouncement(msgs::NodeAnnouncement),
+	ChannelUpdate(msgs::ChannelUpdate),
+	/// A message that could not be decoded because its type is unknown.
+	Unknown(MessageType),
+}
+
+/// A number identifying a message to determine how it is encoded on the wire.
+#[derive(Clone, Copy)]
+pub struct MessageType {
+	number: u16,
+}
+
+impl Message {
+	/// Returns the type that was used to decode the message payload.
+	pub fn r#type(&self) -> MessageType {
```

Comment:
Hmm, never seen this syntax before (and 1.22 seems unhappy about it?)

Reply:
You know, after some further deliberation, I think we may wanna try to look for other words. Raw identifiers look more awkward than, say, something like type_id or get_type or message_type for the method name.

Reply:
I'm fine with something like `type_id` or `type_num`. I'd like to avoid using `message_type` since it makes the call site look awkward as `message.message_id()`.

Reply:
And `get_type` is against the prevailing Rust conventions.

https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

Reply:
Reworked all commits to use `type_id` instead.